pyblish_lite/control.py

```diff
@@ -58,40 +66,107 @@ def __init__(self, parent=None):
             "ordersWithError": set()
         }
 
-    def reset(self):
-        """Discover plug-ins and run collection"""
+    def prepare_for_reset(self):
+        self.validated = False
+        self.extracted = False
+        self.publishing = False
+
         self.context = pyblish.api.Context()
-        self.plugins = pyblish.api.discover()
+        self.context.data["optional"] = False
+        self.context.data["publish"] = True
 
-        self.was_discovered.emit()
+        self.context.data["label"] = 'Context'
+        self.context.data["name"] = 'Context'
+        self.context.data["_type"] = 'context'
 
-        self.pair_generator = None
-        self.current_pair = (None, None)
+        self.context.data["icon"] = 'book'
+
+        self.context.data["_has_succeeded"] = False
+        self.context.data["_has_failed"] = False
+        self.context.data["_is_idle"] = True
+
+        self.context.data["__families__"] = ('__context__',)
+
+    def reset(self):
+        """
+        Discover plug-ins and run collection
+        - prepare_for_reset should be run before
+            - was split because of `Context` in instance list
+            - reset logic should be used only in window.py in `reset` method
+        """
+        # This is just backup
+        if self.context:
+            # - Context probably won't be shown in instance list
+            self.prepare_for_reset()
+
+        # Load collectors
+        self.load_plugins()
         self.current_error = None
+        # Process collectors load rest of plugins with collected instances
+        self.collect()
 
         self.processing = {
             "nextOrder": None,
             "ordersWithError": set()
         }
 
-        self._load()
-        self._run(until=pyblish.api.CollectorOrder,
-                  on_finished=self.was_reset.emit)
+    def load_plugins(self):
+        self.test = pyblish.logic.registered_test()
+        self.processing = {
+            "nextOrder": None,
+            "ordersWithError": set()
+        }
 
-    def validate(self):
-        self._run(until=pyblish.api.ValidatorOrder,
-                  on_finished=self.on_validated)
+        collectors = []
+        validators = []
+        extractors = []
+        conforms = []
+        plugins = pyblish.api.discover()
+
+        targets = pyblish.logic.registered_targets() or ["default"]
+        plugins = pyblish.logic.plugins_by_targets(plugins, targets)
+
+        for plugin in plugins:
+            if plugin.order < (pyblish.api.CollectorOrder + 0.5):
+                collectors.append(plugin)
+            elif plugin.order < (pyblish.api.ValidatorOrder + 0.5):
+                validators.append(plugin)
+            elif plugin.order < (pyblish.api.ExtractorOrder + 0.5):
+                extractors.append(plugin)
+            else:
+                conforms.append(plugin)
+
+        self.plugins = {
```

Comment:
Don't pre-sort these. It's too hard-coded. Look at their `order` to figure out what they are. Remember plug-ins don't have any pre-determined ranges. They are all along one linear scale, running one after the other. CVEI is a mental model, not a technical one.

Reply:
Described above. Don't know how to keep it work properly.